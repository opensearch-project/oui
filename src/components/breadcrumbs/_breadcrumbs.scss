/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/**
 * 1. Add vertical space between breadcrumbs,
 *    but make sure the whole breadcrumb set doesn't add space below itself
 */

.ouiBreadcrumbs {
  @include ouiFontSizeS;
  margin-bottom: -$ouiSizeXS; /* 1 */
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  min-width: 0; // Ensure it shrinks if the window is narrow
}

.ouiBreadcrumb {
  display: inline-block;
  // TODO: remove important: https://github.com/opensearch-project/oui/issues/376
  color: $ouiBreadcrumbInactiveTextColor !important; // sass-lint:disable-line no-important

  &:hover {
    color: $ouiBreadCrumbHoverColor !important; // sass-lint:disable-line no-important
  }
}

.ouiBreadcrumbs:not(.ouiBreadcrumbs__inPopover) .ouiBreadcrumbWrapper--last .ouiBreadcrumb {
  font-weight: $ouiFontWeightMedium;
}

.ouiBreadcrumb--collapsed {
  flex-shrink: 0;
  color: $ouiBreadcrumbCollapsedLink !important; // sass-lint:disable-line no-important
  vertical-align: top !important; // sass-lint:disable-line no-important
}

.ouiBreadcrumb__collapsedLink:hover {
  color: $ouiBreadCrumbHoverColor !important; // sass-lint:disable-line no-important
}

.ouiBreadcrumbs--truncate {
  white-space: nowrap;
  flex-wrap: nowrap;

  .ouiBreadcrumb:not(.ouiBreadcrumb--collapsed) {
    max-width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
    vertical-align: top; // overflow hidden causes misalignment of links and slashes, this fixes that
  }

  .ouiBreadcrumbWrapper:not(.ouiBreadcrumbWrapper--collapsed) {
    max-width: $ouiBreadcrumbTruncateWidth;
    overflow: hidden;
    text-overflow: ellipsis;

    &.ouiBreadcrumbWrapper--last {
      max-width: none;
    }
  }
}

.ouiBreadcrumb--truncate {
  @include ouiTextTruncate;
  max-width: 100%;
  text-align: center;
  vertical-align: top; // overflow hidden causes misalignment of links and slashes, this fixes that
}

.ouiBreadcrumbWrapper--truncate {
  max-width: $ouiBreadcrumbTruncateWidth;
}

.ouiBreadcrumbWrapper {
  position: relative;
  z-index: 0;
  padding: $ouiSizeXS - 2.5 $ouiSizeL - $ouiSizeXS;
  padding-right: $ouiSizeL - $ouiSizeXS + calc($ouiBreadcrumbSpacing / 2);

  &::before {
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: calc($ouiBreadcrumbSpacing / 2);
    right: calc($ouiBreadcrumbSpacing / 2);
    z-index: -1;

    transform: skewX(-20deg);
    border-radius: $ouiSizeXS;
  }

  &:not(.ouiBreadcrumbWrapper--last)::before {
    background-color: $ouiBreadcrumbInactiveBackground;
  }

  &.ouiBreadcrumbWrapper--last::before {
    background-color: $ouiBreadcrumbActiveBackground;
  }

  &:not(.ouiBreadcrumbWrapper--first) {
    margin-bottom: $ouiSizeXS; /* 1 */
    padding-left: $ouiSizeL - $ouiSizeXS + calc($ouiBreadcrumbSpacing / 2);
  }

  // This targets the last breadcrumb wrapper and sets a different text color for the breadcrumb inside it.
  &.ouiBreadcrumbWrapper--last .ouiBreadcrumb {
    // TODO: remove important: https://github.com/opensearch-project/oui/issues/376
    color: inherit !important; // sass-lint:disable-line no-important

    &:hover {
      color: inherit !important; // sass-lint:disable-line no-important
    }
  }
}

.ouiBreadcrumbWall {
  background-image: linear-gradient(to right, $ouiBreadcrumbInactiveBackground 0 $ouiSizeM, transparent $ouiSizeM);
  border-radius: $ouiSizeXS;
  overflow: hidden;
  margin-bottom: $ouiSizeXS; /* 1 */
}

.ouiBreadcrumbWall--single {
  background-image: linear-gradient(to right, $ouiBreadcrumbActiveBackground 0 $ouiSizeM, transparent $ouiSizeM);
}

.euiLink.euiBreadcrumb {
  line-height: inherit;
  font-weight: inherit;
}
