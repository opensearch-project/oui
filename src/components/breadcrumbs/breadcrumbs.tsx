/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  Fragment,
  FunctionComponent,
  MouseEventHandler,
  ReactNode,
  useEffect,
  useState,
} from 'react';
import classNames from 'classnames';

import { CommonProps } from '../common';
import { OuiI18n } from '../i18n';
import { OuiInnerText } from '../inner_text';
import { OuiLink } from '../link';
import { OuiPopover } from '../popover';
import { OuiIcon } from '../icon';
import { throttle } from '../../services';
import { OuiBreakpointSize, getBreakpoint } from '../../services/breakpoint';

export type OuiBreadcrumbResponsiveMaxCount = {
  /**
   * Any of the following keys are allowed: `'xs' | 's' | 'm' | 'l' | 'xl'`
   * Omitting a key will display all breadcrumbs at that breakpoint
   */
  [key in OuiBreakpointSize]?: number;
};

export type OuiBreadcrumb = CommonProps & {
  /**
   * Visible label of the breadcrumb
   */
  text: ReactNode;
  href?: string;
  onClick?: MouseEventHandler<HTMLAnchorElement>;
  /**
   * Force a max-width on the breadcrumb text
   */
  truncate?: boolean;
};

export type OuiBreadcrumbsProps = CommonProps & {
  /**
   * Hides extra (above the max) breadcrumbs under a collapsed item as the window gets smaller.
   * Pass a custom #OuiBreadcrumbResponsiveMaxCount object to change the number of breadcrumbs to show at the particular breakpoints.
   * Omitting or passing a `0` value will show all breadcrumbs.
   *
   * Pass `false` to turn this behavior off.
   *
   * Default: `{ xs: 1, s: 2, m: 4 }`
   */
  responsive?: boolean | OuiBreadcrumbResponsiveMaxCount;

  /**
   * Forces all breadcrumbs to single line and
   * truncates each breadcrumb to a particular width,
   * except for the last item
   */
  truncate?: boolean;

  /**
   * Collapses the inner items past the maximum set here
   * into a single ellipses item
   */
  max?: number | null;

  /**
   * The array of individual #OuiBreadcrumb items
   */
  breadcrumbs: OuiBreadcrumb[];
};

const responsiveDefault: OuiBreadcrumbResponsiveMaxCount = {
  xs: 1,
  s: 2,
  m: 4,
};

const limitBreadcrumbs = (
  breadcrumbs: ReactNode[],
  max: number,
  allBreadcrumbs: OuiBreadcrumb[]
) => {
  const breadcrumbsAtStart = [];
  const breadcrumbsAtEnd = [];
  const limit = Math.min(max, breadcrumbs.length);
  const start = Math.floor(limit / 2);
  const overflowBreadcrumbs = allBreadcrumbs.slice(
    start,
    start + breadcrumbs.length - limit
  );

  for (let i = 0; i < limit; i++) {
    // We'll alternate with displaying breadcrumbs at the end and at the start, but be biased
    // towards breadcrumbs the end so that if max is an odd number, we'll have one more
    // breadcrumb visible at the end than at the beginning.
    const isEven = i % 2 === 0;

    // We're picking breadcrumbs from the front AND the back, so we treat each iteration as a
    // half-iteration.
    const normalizedIndex = Math.floor(i * 0.5);
    const indexOfBreadcrumb = isEven
      ? breadcrumbs.length - 1 - normalizedIndex
      : normalizedIndex;
    const breadcrumb = breadcrumbs[indexOfBreadcrumb];

    if (isEven) {
      breadcrumbsAtEnd.unshift(breadcrumb);
    } else {
      breadcrumbsAtStart.push(breadcrumb);
    }
  }

  const OuiBreadcrumbCollapsed = () => {
    const [isPopoverOpen, setIsPopoverOpen] = useState(false);

    const ellipsisButton = (
      <OuiI18n
        token="ouiBreadcrumbs.collapsedBadge.ariaLabel"
        default="Show collapsed breadcrumbs">
        {(ariaLabel: string) => (
          <OuiLink
            className="ouiBreadcrumb__collapsedLink"
            color="subdued"
            aria-label={ariaLabel}
            title={ariaLabel}
            onClick={() => setIsPopoverOpen(!isPopoverOpen)}>
            &hellip; <OuiIcon type="arrowDown" size="s" />
          </OuiLink>
        )}
      </OuiI18n>
    );

    return (
      <Fragment>
        <div className="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed">
          <OuiPopover
            className="ouiBreadcrumb ouiBreadcrumb--collapsed"
            button={ellipsisButton}
            isOpen={isPopoverOpen}
            closePopover={() => setIsPopoverOpen(false)}>
            <OuiBreadcrumbs
              className="ouiBreadcrumbs__inPopover"
              breadcrumbs={overflowBreadcrumbs}
              responsive={false}
              truncate={false}
              max={0}
            />
          </OuiPopover>
        </div>
      </Fragment>
    );
  };

  if (max < breadcrumbs.length) {
    breadcrumbsAtStart.push(<OuiBreadcrumbCollapsed key="collapsed" />);
  }

  return [...breadcrumbsAtStart, ...breadcrumbsAtEnd];
};

export const OuiBreadcrumbs: FunctionComponent<OuiBreadcrumbsProps> = ({
  breadcrumbs,
  className,
  responsive = responsiveDefault,
  truncate = true,
  max = 5,
  ...rest
}) => {
  const [currentBreakpoint, setCurrentBreakpoint] = useState(
    getBreakpoint(typeof window === 'undefined' ? -Infinity : window.innerWidth)
  );

  const functionToCallOnWindowResize = throttle(() => {
    const newBreakpoint = getBreakpoint(window.innerWidth);
    if (newBreakpoint !== currentBreakpoint) {
      setCurrentBreakpoint(newBreakpoint);
    }
    // reacts every 50ms to resize changes and always gets the final update
  }, 50);

  // Add window resize handlers
  useEffect(() => {
    window.addEventListener('resize', functionToCallOnWindowResize);

    return () => {
      window.removeEventListener('resize', functionToCallOnWindowResize);
    };
  }, [responsive, functionToCallOnWindowResize]);

  const breadcrumbElements = breadcrumbs.map((breadcrumb, index) => {
    const {
      text,
      href,
      onClick,
      truncate,
      className: breadcrumbClassName,
      ...breadcrumbRest
    } = breadcrumb;

    const isFirstBreadcrumb = index === 0;
    const isLastBreadcrumb = index === breadcrumbs.length - 1;

    const breadcrumbWrapperClasses = classNames('ouiBreadcrumbWrapper', {
      'ouiBreadcrumbWrapper--first': isFirstBreadcrumb,
      'ouiBreadcrumbWrapper--last': isLastBreadcrumb,
      'ouiBreadcrumbWrapper--truncate': truncate,
    });

    const breadcrumbClasses = classNames('ouiBreadcrumb', breadcrumbClassName, {
      'ouiBreadcrumb--last': isLastBreadcrumb,
      'ouiBreadcrumb--truncate': truncate,
    });

    const link =
      !href && !onClick ? (
        <OuiInnerText>
          {(ref, innerText) => (
            <span
              ref={ref}
              className={breadcrumbClasses}
              title={innerText}
              aria-current={isLastBreadcrumb ? 'page' : 'false'}
              {...breadcrumbRest}>
              {text}
            </span>
          )}
        </OuiInnerText>
      ) : (
        <OuiInnerText>
          {(ref, innerText) => (
            <OuiLink
              ref={ref}
              color={isLastBreadcrumb ? 'text' : 'subdued'}
              onClick={onClick}
              href={href}
              className={breadcrumbClasses}
              title={innerText}
              {...breadcrumbRest}>
              {text}
            </OuiLink>
          )}
        </OuiInnerText>
      );

    const breadcrumbWallClasses = classNames('ouiBreadcrumbWall', {
      'ouiBreadcrumbWall--single': isFirstBreadcrumb && isLastBreadcrumb,
    });

    const wrapper = <div className={breadcrumbWrapperClasses}>{link}</div>;
    const wall = isFirstBreadcrumb ? (
      <div className={breadcrumbWallClasses}>{wrapper}</div>
    ) : (
      wrapper
    );

    return <Fragment key={index}>{wall}</Fragment>;
  });

  // Use the default object if they simply passed `true` for responsive
  const responsiveObject =
    typeof responsive === 'object' ? responsive : responsiveDefault;

  // The max property collapses any breadcrumbs past the max quantity.
  // This is the same behavior we want for responsiveness.
  // So calculate the max value based on the combination of `max` and `responsive`

  // First, calculate the responsive max value
  const responsiveMax =
    responsive && responsiveObject[currentBreakpoint as OuiBreakpointSize]
      ? responsiveObject[currentBreakpoint as OuiBreakpointSize]
      : null;

  // Second, if both max and responsiveMax are set, use the smaller of the two. Otherwise, use the one that is set.
  const calculatedMax: OuiBreadcrumbsProps['max'] =
    max && responsiveMax ? Math.min(max, responsiveMax) : max || responsiveMax;

  const limitedBreadcrumbs = calculatedMax
    ? limitBreadcrumbs(breadcrumbElements, calculatedMax, breadcrumbs)
    : breadcrumbElements;

  const classes = classNames('ouiBreadcrumbs', className, {
    'ouiBreadcrumbs--truncate': truncate,
  });

  return (
    <nav aria-label="breadcrumb" className={classes} {...rest}>
      {limitedBreadcrumbs}
    </nav>
  );
};
