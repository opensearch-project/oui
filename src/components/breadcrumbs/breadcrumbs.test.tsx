/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { requiredProps } from '../../test';

import { OuiBreadcrumbs } from './breadcrumbs';

const breadcrumbs = [
  {
    text: 'Animals',
    href: '#',
    onClick: (e: React.MouseEvent) => {
      e.preventDefault();
      console.log('You clicked Animals');
    },
    'data-test-subj': 'breadcrumbsAnimals',
    className: 'customClass',
  },
  {
    text: 'Metazoans',
  },
  {
    text: 'Chordates',
  },
  {
    text:
      'Nebulosa subspecies is also a real mouthful, especially for creatures without mouths',
    truncate: true,
  },
  {
    text: 'Tetrapods',
  },
  {
    text: 'Reptiles',
    onClick: (e: React.MouseEvent) => {
      e.preventDefault();
      console.log('You clicked Reptiles');
    },
  },
  {
    text: 'Boa constrictor',
    href: '#',
    truncate: true,
  },
  {
    text: 'Edit',
  },
];

describe('OuiBreadcrumbs', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiBreadcrumbs {...requiredProps} breadcrumbs={breadcrumbs} />
    );

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    describe('responsive', () => {
      test('is rendered', () => {
        const { container } = render(
          <OuiBreadcrumbs breadcrumbs={breadcrumbs} responsive />
        );
        expect(container).toMatchSnapshot();
      });

      test('is rendered as false', () => {
        const { container } = render(
          <OuiBreadcrumbs breadcrumbs={breadcrumbs} responsive={false} />
        );
        expect(container).toMatchSnapshot();
      });

      test('is rendered with custom breakpoints', () => {
        const { container } = render(
          <OuiBreadcrumbs
            breadcrumbs={breadcrumbs}
            responsive={{ xs: 1, s: 1, m: 1, l: 1, xl: 1 }}
          />
        );
        expect(container).toMatchSnapshot();
      });
    });

    describe('truncate as false', () => {
      test('is rendered', () => {
        const { container } = render(
          <OuiBreadcrumbs breadcrumbs={breadcrumbs} truncate={false} />
        );
        expect(container).toMatchSnapshot();
      });
    });

    describe('max', () => {
      test('renders 1 item', () => {
        const { container } = render(
          <OuiBreadcrumbs breadcrumbs={breadcrumbs} max={1} />
        );
        expect(container).toMatchSnapshot();
      });

      test('renders all items with null', () => {
        const { container } = render(
          <OuiBreadcrumbs breadcrumbs={breadcrumbs} max={null} />
        );
        expect(container).toMatchSnapshot();
      });

      test("doesn't break when max exceeds the number of breadcrumbs", () => {
        const { container } = render(
          <OuiBreadcrumbs breadcrumbs={breadcrumbs} max={20} />
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
