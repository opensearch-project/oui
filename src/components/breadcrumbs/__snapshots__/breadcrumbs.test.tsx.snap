// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiBreadcrumbs is rendered 1`] = `
<nav
  aria-label="aria-label"
  class="ouiBreadcrumbs testClass1 testClass2 ouiBreadcrumbs--truncate"
  data-test-subj="test subject string"
>
  <div
    class="ouiBreadcrumbWall"
  >
    <div
      class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--first"
    >
      <a
        class="ouiLink ouiLink--subdued ouiBreadcrumb customClass"
        data-test-subj="breadcrumbsAnimals"
        href="#"
        rel="noreferrer"
      >
        Animals
      </a>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Metazoans
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed"
  >
    <div
      class="ouiPopover ouiPopover--anchorDownCenter ouiBreadcrumb ouiBreadcrumb--collapsed"
    >
      <div
        class="ouiPopover__anchor"
      >
        <button
          aria-label="Show collapsed breadcrumbs"
          class="ouiLink ouiLink--subdued ouiBreadcrumb__collapsedLink"
          title="Show collapsed breadcrumbs"
          type="button"
        >
          … 
          <span
            data-ouiicon-type="arrowDown"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <button
      class="ouiLink ouiLink--subdued ouiBreadcrumb"
      type="button"
    >
      Reptiles
    </button>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <a
      class="ouiLink ouiLink--subdued ouiBreadcrumb ouiBreadcrumb--truncate"
      href="#"
      rel="noreferrer"
    >
      Boa constrictor
    </a>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props max doesn't break when max exceeds the number of breadcrumbs 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs ouiBreadcrumbs--truncate"
>
  <div
    class="ouiBreadcrumbWall"
  >
    <div
      class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--first"
    >
      <a
        class="ouiLink ouiLink--subdued ouiBreadcrumb customClass"
        data-test-subj="breadcrumbsAnimals"
        href="#"
        rel="noreferrer"
      >
        Animals
      </a>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Metazoans
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Chordates
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb ouiBreadcrumb--truncate"
    >
      Nebulosa subspecies is also a real mouthful, especially for creatures without mouths
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Tetrapods
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <button
      class="ouiLink ouiLink--subdued ouiBreadcrumb"
      type="button"
    >
      Reptiles
    </button>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <a
      class="ouiLink ouiLink--subdued ouiBreadcrumb ouiBreadcrumb--truncate"
      href="#"
      rel="noreferrer"
    >
      Boa constrictor
    </a>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props max renders 1 item 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs ouiBreadcrumbs--truncate"
>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed"
  >
    <div
      class="ouiPopover ouiPopover--anchorDownCenter ouiBreadcrumb ouiBreadcrumb--collapsed"
    >
      <div
        class="ouiPopover__anchor"
      >
        <button
          aria-label="Show collapsed breadcrumbs"
          class="ouiLink ouiLink--subdued ouiBreadcrumb__collapsedLink"
          title="Show collapsed breadcrumbs"
          type="button"
        >
          … 
          <span
            data-ouiicon-type="arrowDown"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props max renders all items with null 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs ouiBreadcrumbs--truncate"
>
  <div
    class="ouiBreadcrumbWall"
  >
    <div
      class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--first"
    >
      <a
        class="ouiLink ouiLink--subdued ouiBreadcrumb customClass"
        data-test-subj="breadcrumbsAnimals"
        href="#"
        rel="noreferrer"
      >
        Animals
      </a>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Metazoans
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Chordates
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb ouiBreadcrumb--truncate"
    >
      Nebulosa subspecies is also a real mouthful, especially for creatures without mouths
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Tetrapods
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <button
      class="ouiLink ouiLink--subdued ouiBreadcrumb"
      type="button"
    >
      Reptiles
    </button>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <a
      class="ouiLink ouiLink--subdued ouiBreadcrumb ouiBreadcrumb--truncate"
      href="#"
      rel="noreferrer"
    >
      Boa constrictor
    </a>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props responsive is rendered 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs ouiBreadcrumbs--truncate"
>
  <div
    class="ouiBreadcrumbWall"
  >
    <div
      class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--first"
    >
      <a
        class="ouiLink ouiLink--subdued ouiBreadcrumb customClass"
        data-test-subj="breadcrumbsAnimals"
        href="#"
        rel="noreferrer"
      >
        Animals
      </a>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Metazoans
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed"
  >
    <div
      class="ouiPopover ouiPopover--anchorDownCenter ouiBreadcrumb ouiBreadcrumb--collapsed"
    >
      <div
        class="ouiPopover__anchor"
      >
        <button
          aria-label="Show collapsed breadcrumbs"
          class="ouiLink ouiLink--subdued ouiBreadcrumb__collapsedLink"
          title="Show collapsed breadcrumbs"
          type="button"
        >
          … 
          <span
            data-ouiicon-type="arrowDown"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <button
      class="ouiLink ouiLink--subdued ouiBreadcrumb"
      type="button"
    >
      Reptiles
    </button>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <a
      class="ouiLink ouiLink--subdued ouiBreadcrumb ouiBreadcrumb--truncate"
      href="#"
      rel="noreferrer"
    >
      Boa constrictor
    </a>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props responsive is rendered as false 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs ouiBreadcrumbs--truncate"
>
  <div
    class="ouiBreadcrumbWall"
  >
    <div
      class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--first"
    >
      <a
        class="ouiLink ouiLink--subdued ouiBreadcrumb customClass"
        data-test-subj="breadcrumbsAnimals"
        href="#"
        rel="noreferrer"
      >
        Animals
      </a>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Metazoans
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed"
  >
    <div
      class="ouiPopover ouiPopover--anchorDownCenter ouiBreadcrumb ouiBreadcrumb--collapsed"
    >
      <div
        class="ouiPopover__anchor"
      >
        <button
          aria-label="Show collapsed breadcrumbs"
          class="ouiLink ouiLink--subdued ouiBreadcrumb__collapsedLink"
          title="Show collapsed breadcrumbs"
          type="button"
        >
          … 
          <span
            data-ouiicon-type="arrowDown"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <button
      class="ouiLink ouiLink--subdued ouiBreadcrumb"
      type="button"
    >
      Reptiles
    </button>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <a
      class="ouiLink ouiLink--subdued ouiBreadcrumb ouiBreadcrumb--truncate"
      href="#"
      rel="noreferrer"
    >
      Boa constrictor
    </a>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props responsive is rendered with custom breakpoints 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs ouiBreadcrumbs--truncate"
>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed"
  >
    <div
      class="ouiPopover ouiPopover--anchorDownCenter ouiBreadcrumb ouiBreadcrumb--collapsed"
    >
      <div
        class="ouiPopover__anchor"
      >
        <button
          aria-label="Show collapsed breadcrumbs"
          class="ouiLink ouiLink--subdued ouiBreadcrumb__collapsedLink"
          title="Show collapsed breadcrumbs"
          type="button"
        >
          … 
          <span
            data-ouiicon-type="arrowDown"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;

exports[`OuiBreadcrumbs props truncate as false is rendered 1`] = `
<nav
  aria-label="breadcrumb"
  class="ouiBreadcrumbs"
>
  <div
    class="ouiBreadcrumbWall"
  >
    <div
      class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--first"
    >
      <a
        class="ouiLink ouiLink--subdued ouiBreadcrumb customClass"
        data-test-subj="breadcrumbsAnimals"
        href="#"
        rel="noreferrer"
      >
        Animals
      </a>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <span
      aria-current="false"
      class="ouiBreadcrumb"
    >
      Metazoans
    </span>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--collapsed"
  >
    <div
      class="ouiPopover ouiPopover--anchorDownCenter ouiBreadcrumb ouiBreadcrumb--collapsed"
    >
      <div
        class="ouiPopover__anchor"
      >
        <button
          aria-label="Show collapsed breadcrumbs"
          class="ouiLink ouiLink--subdued ouiBreadcrumb__collapsedLink"
          title="Show collapsed breadcrumbs"
          type="button"
        >
          … 
          <span
            data-ouiicon-type="arrowDown"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="ouiBreadcrumbWrapper"
  >
    <button
      class="ouiLink ouiLink--subdued ouiBreadcrumb"
      type="button"
    >
      Reptiles
    </button>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--truncate"
  >
    <a
      class="ouiLink ouiLink--subdued ouiBreadcrumb ouiBreadcrumb--truncate"
      href="#"
      rel="noreferrer"
    >
      Boa constrictor
    </a>
  </div>
  <div
    class="ouiBreadcrumbWrapper ouiBreadcrumbWrapper--last"
  >
    <span
      aria-current="page"
      class="ouiBreadcrumb ouiBreadcrumb--last"
    >
      Edit
    </span>
  </div>
</nav>
`;
